src/oslogin_utils.cc

```diff
@@ -434,95 +466,122 @@ bool ParseJsonToGroups(const string& json, std::vector<Group>* result) {
-      return false;
+      goto cleanup;
     }
 
     Group g;
     g.gid = json_object_get_int64(gid);
 
     // get_int64 will confusingly return 0 if the string can't be converted to
     // an integer. We can't rely on type check as it may be a string in the API.
+    // Also 0 is invalid because it creates a 'root group'.
     if (g.gid == 0) {
-      return false;
+      goto cleanup;
     }
+
     g.name = json_object_get_string(name);
     if (g.name == "") {
-      return false;
+      goto cleanup;
     }
 
     result->push_back(g);
   }
-  return true;
+  ret = true;
+
+cleanup:
+  json_object_put(root);
+  return ret;
 }
 
-bool ParseJsonToGroup(const string& json, struct group* result, BufferManager* buf, int* errnop) {
+bool ParseJsonToGroup(const string& json, struct group* result, BufferManager*
+                      buf, int* errnop) {
+  *errnop = EINVAL;
+  int gr_gid = 65535;
+
   json_object* group = NULL;
   group = json_tokener_parse(json.c_str());
-  if (group== NULL) {
-    *errnop = EINVAL;
+  if (group == NULL) {
     return false;
   }
 
+  bool ret = false;
+
   json_object* gid;
   if (!json_object_object_get_ex(group, "gid", &gid)) {
-    *errnop = EINVAL;
-    return false;
+    goto cleanup;
   }
 
   json_object* name;
   if (!json_object_object_get_ex(group, "name", &name)) {
-    *errnop = EINVAL;
-    return false;
+    goto cleanup;
+  }
+
+  if ((gr_gid = json_object_get_int64(gid)) == 0) {
+    goto cleanup;
   }
 
-  result->gr_gid = json_object_get_int64(gid);
-  // TODO ValidateGroup
-  buf->AppendString("", &result->gr_passwd, errnop);
-  return buf->AppendString((char*)json_object_get_string(name), &result->gr_name, errnop);
+  result->gr_gid = gr_gid;
+  if (!buf->AppendString("", &result->gr_passwd, errnop))
+    goto cleanup;
+  if (!buf->AppendString((char*)json_object_get_string(name), &result->gr_name,
+                         errnop))
+    goto cleanup;
+
+  *errnop = 0;
+  ret = true;
+
+cleanup:
+  json_object_put(group);
+  return ret;
 }
 
 std::vector<string> ParseJsonToSshKeys(const string& json) {
   std::vector<string> result;
+  json_object* ssh_public_keys = NULL;
+
   json_object* root = NULL;
   root = json_tokener_parse(json.c_str());
   if (root == NULL) {
     return result;
   }
+
   // Locate the sshPublicKeys object.
   json_object* login_profiles = NULL;
   if (!json_object_object_get_ex(root, "loginProfiles", &login_profiles)) {
-    return result;
+    goto cleanup;
   }
   if (json_object_get_type(login_profiles) != json_type_array) {
-    return result;
+    goto cleanup;
   }
   login_profiles = json_object_array_get_idx(login_profiles, 0);
 
-  json_object* ssh_public_keys = NULL;
   if (!json_object_object_get_ex(login_profiles, "sshPublicKeys", &ssh_public_keys)) {
-    return result;
+    goto cleanup;
   }
 
   if (json_object_get_type(ssh_public_keys) != json_type_object) {
-    return result;
+    goto cleanup;
   }
+  {
```

Comment:
this additional brace indicates the scope of the variables initialized within; otherwise g++ will complain the goto steps over initializers. this is a c++ism. same reason we moved a lot of declarations/initializations to top of function.

Reply:
Thank you for explaining this! Can you actually add it as a comment in the code?